### DISCUSSION TITLE
add option to prevent menu item icon_color from being discolored when selected

### DISCUSSION THREAD

**Discussion by Mariusthvdb** [elchininet/custom-sidebar#202] [Ideas]
we can color the icons of the menu items, and even when hovering, the icons keep their color if so set, either find, or dynamic via a template

However, when a menu item is selected, the colorization of the item breaks the icon_color:

<img width="263" alt="Scherm­afbeelding 2024-11-13 om 15 24 34" src="https://github.com/user-attachments/assets/0c67d131-e731-4ee2-8221-add8fe40c698">
<img width="255" alt="Scherm­afbeelding 2024-11-13 om 15 24 42" src="https://github.com/user-attachments/assets/a26da1fd-0f46-433d-b4e3-5a7504106c36">
<img width="249" alt="Scherm­afbeelding 2024-11-13 om 15 24 49" src="https://github.com/user-attachments/assets/64008df4-4063-4b7c-ad11-8b67945022b9">

It would be very welcome to keep have an option and keep the icons colorized

please consider adding that?
thx.

**Comment by elchininet** [comment]
There is an `icon_color_selected` option, when this option is not set, the icon gets the color defined in `--sidebar-selected-icon-color` variable. If you want to keep the same color just set the `icon_color` and the `icon_color_selected` options with the same value.

**Reply by Mariusthvdb** [reply]
but then I have to write those templates twice...

could some plugin variable be created, so we can add that (if one uses a template on the `icon_color`?

maybe simply add 
```
icon_color_selected: var(--custom-sidebar-icon-color)
```

?
Ofc this should then be the actual item icon_color…

**Reply by elchininet** [reply]
The plugin overrides core styles, so when the item is selected, instead of `var(--sidebar-selected-icon-color)` it uses `var(--custom-sidebar-selected-icon-color, var(--sidebar-selected-icon-color))`. You can override the plugin styles and have your own override with what you want:

```yaml
styles: |
  :host .iron-selected > paper-icon-item > :is(ha-svg-icon, ha-icon) {
    color: var(--custom-sidebar-icon-color, var(--sidebar-icon-color));
  }
```

**Reply by Mariusthvdb** [reply]
not sure this is what I meant to ask, let me rephrase to be sure.
consider a config fully made inside the plugin (so, no theming or other colors are in play)

for an item, I set
```
icon_color: green
```
and it shows fine, when other items are selected, and even when hovering

however, when selecting the item, that icon_color is no longer green, but gets the default ha `sidebar-icon-color`

My hope is to somehow keep the icon_color I set in the plugin for that item (so, not the generic icon_color) in a variable `--custom-sidebar-item-icon-color`


example:
```
  - item: ui-overzicht
    match: data-panel
    order: 1
    icon_color: >
      [[[ @partial overzicht
          return waarschuwing ? 'var(--alert-color)' : 'var(--ok-color)'; ]]]
    icon_color_selected: var(--custom-sidebar-item-icon-color) # referencing the icon_color of this item above, without having to duplicate it
```
Hope this makes sense

**Reply by elchininet** [reply]
Let me explain better:

>and it shows fine, when other items are selected, and even when hovering

Home Assistant doesn't do anything with hovering, hovering is a custom thing that you introduced so that is the why it doesn't get affected.

Now, getting another color when the item is selected is a Home Assistant feature, that happens even if you don't have `custom-sidebar` installed, Home Assistant sets `var(--sidebar-selected-icon-color)` when an item is selected. What the plugin does is just to add a custom variable before than one to override core: `var(--custom-sidebar-selected-icon-color, var(--sidebar-selected-icon-color))`.

If you want to to override that behaviour from Home Assistant do your own override. The style example that I gave you does what you want **for every icon in the sidebar**, it gets the `icon_color` (if you have one in the item, if not the generic one) and if you don't have one in the item neither a generic one, then it gets the color from the theme.

>My hope is to somehow keep the icon_color I set in the plugin for that item (so, not the generic icon_color) in a variable --custom-sidebar-item-icon-color

`icon_color` inside an item overrides the generic `icon_color`, so if you set it in an item, the variable `--sidebar-icon-color` will contain that color. The style code that I posted before does exactly that.

But if you want to do it on every item, instead of:

```yaml
  icon_color_selected: var(--custom-sidebar-item-icon-color) # referencing the icon_color of this item above, without having to duplicate it
```

You should use:

```yaml
  icon_color_selected: var(--custom-sidebar-icon-color) # referencing the icon_color of this item above, without having to duplicate it
```

Example with code

```yaml
icon_color: red
order:
  - item: ui-overzicht
    order: 1
    icon_color: green
    icon_color_selected: var(--custom-sidebar-icon-color) # --custom-sidebar-icon-color is green
  - item: another-item
    icon_color_selected: var(--custom-sidebar-icon-color) # --custom-sidebar-icon-color is red
```

But as I mentioned, you will get the same behaviour with the style override that I posted before.

**Reply by Mariusthvdb** [reply]
the per item option
```
icon_color_selected: var(--custom-sidebar-icon-color)
```

seems to almost be perfect, not sure what I did wrong yesterday, because I could have sworn to have tested that.... it does however also set the background of the selected item to the icon_color, and that is not what I want..

not selected:

<img width="257" alt="Scherm­afbeelding 2024-11-14 om 09 48 26" src="https://github.com/user-attachments/assets/07215af0-64f3-46e6-bc35-840cc27cbb2f">

selected:

<img width="266" alt="Scherm­afbeelding 2024-11-14 om 09 48 33" src="https://github.com/user-attachments/assets/bfa61bc5-e158-4265-abb2-419526912fef">

setting `icon_color_selected: var(--custom-sidebar-icon-color)` as global option does Not keep the color of the dynamically set icon, (and also does Not color the background)

as a matter of fact, I can spot no effect at all of setting that option.

**Reply by elchininet** [reply]
>it does however also set the background of the selected item to the icon_color, and that is not what I want..

That is also a Home Assistant feature, the background of the selected item is by default `var(--sidebar-selected-icon-color)` with an opacity of `0.12`. If you set `icon_color_selected` then `custom-sidebar` overrides that value and the background will change. That is why the `selection_color` and `selection_opacity` options exist. As you want to keep the background of all the selected items with `var(--sidebar-selected-icon-color)` just set it as a global option:

```yaml
selection_color: var(--sidebar-selected-icon-color)
```

>setting icon_color_selected: var(--custom-sidebar-icon-color) as global option does Not keep the color of the dynamically set icon

That should not happen, most probably there is a rule from the same `custom-sidebar` that is overriding that. I need to check it, but if it is not working most probably it is a bug.

**Reply by Mariusthvdb** [reply]
almost!

`selection_color: var(--sidebar-selected-icon-color)` makes the selection color back to 'normal'

<img width="249" alt="Scherm­afbeelding 2024-11-14 om 10 19 38" src="https://github.com/user-attachments/assets/67351f34-37c8-4520-ab36-5318a7dfadd9">
<img width="252" alt="Scherm­afbeelding 2024-11-14 om 10 19 53" src="https://github.com/user-attachments/assets/2f5b56ad-4858-419d-94f8-aa83c0195aa4">

but as you can see still overrides the `info_color` and `text_color` with that same default


they should show as:
<img width="249" alt="Scherm­afbeelding 2024-11-14 om 10 21 51" src="https://github.com/user-attachments/assets/57119d5c-674e-452c-aa22-8d42223cfdf7">
<img width="251" alt="Scherm­afbeelding 2024-11-14 om 10 21 45" src="https://github.com/user-attachments/assets/1411b424-b71e-4abc-8c26-6b9b88454914">

**Reply by elchininet** [reply]
>but as you can see still overrides the info_color with that same default

That is also a Home Assistant feature (by the way, it should happen even if you don't set that option).

When the item is selected, Home Assistant sets the color of the text as `var(--sidebar-selected-text-color)`. That is why there are `text_color_selected` and `info_color_selected` options.

**Reply by Mariusthvdb** [reply]
haha, nice, yes confirmed
```
    icon_color_selected: var(--custom-sidebar-icon-color)
    text_color_selected: var(--custom-sidebar-text-color)
    info_color_selected: var(--custom-sidebar-info-color)
```

on the individual items keeps the colors in tact

<img width="252" alt="Scherm­afbeelding 2024-11-14 om 10 39 13" src="https://github.com/user-attachments/assets/7c00c0a3-8289-4042-9a52-c589414029d0">
<img width="255" alt="Scherm­afbeelding 2024-11-14 om 10 39 21" src="https://github.com/user-attachments/assets/733be000-3761-4372-b41a-df0b62e1ee59">

man this is becoming quite the config..... now if we could make that happen in the main config, that would be much easier, let me see if these last options do work


nope, they dont, we need to set them on each individual item.

fixing that by creating a yaml anchor

```
    <<: &keep
      icon_color_selected: var(--custom-sidebar-icon-color)
      text_color_selected: var(--custom-sidebar-text-color)
      info_color_selected: var(--custom-sidebar-info-color)
```

and pasting that in the individual items below
```
    <<: *keep
```

makes it happen

**Reply by elchininet** [reply]
>now if we could make that happen in the main config, that would be much easier, let me see if these last options do work

That should be a bug, no worries, when I have some time I'll look at it and try to fix it. For now you can also use the `styles` option, that should do the work.

**Reply by elchininet** [reply]
@Mariusthvdb,
Try with version `v6.6.1`, it fixes the bug of the global variables not being respected. You can move all the local options.

**Reply by Mariusthvdb** [reply]
confirmed!!

![Nov-14-2024 22-44-01](https://github.com/user-attachments/assets/b2be367e-b487-482c-85a3-6c9b1bf0251d)

**Reply by Mariusthvdb** [reply]
uhm, wait a second, that is hovering.....

it doesnt work when selected..... I am sorry

Ive set this now globally, and took out the anchors on the items, and dont see the colors when selected, just like before

<img width="252" alt="Scherm­afbeelding 2024-11-14 om 22 52 46" src="https://github.com/user-attachments/assets/7b023e1f-7f5a-47ec-88d3-492459969173">

while setting the same on the individual items brings back those colors:

<img width="256" alt="Scherm­afbeelding 2024-11-14 om 22 54 42" src="https://github.com/user-attachments/assets/2e78c7c2-984e-46ae-a3a2-d409c9c536dd">

**Reply by elchininet** [reply]
>it doesnt work when selected..... I am sorry
>
>Ive set this now globally, and took out the anchors on the items, and dont see the colors when selected, jut like before

Post an example of the code that you are using.

**Reply by Mariusthvdb** [reply]
```
##########################################################################################
# Main settings
##########################################################################################
icon_color_selected: var(--custom-sidebar-icon-color)
text_color_selected: var(--custom-sidebar-text-color)
info_color_selected: var(--custom-sidebar-info-color)
```

doesnt work

while doing

```
  - item: ui-overzicht
    match: data-panel
    order: 1
    <<: &keep
      icon_color_selected: var(--custom-sidebar-icon-color)
      text_color_selected: var(--custom-sidebar-text-color)
      info_color_selected: var(--custom-sidebar-info-color)

  - item: ui-develop
    match: data-panel
    order: 5
    <<: *keep
    icon: >
      [[[ @partial systeem
          return update ? 'mdi:package-up' : 'mdi:home-assistant'; ]]]
    icon_color: >
      [[[ @partial systeem
          return update || geheugen > 30 ? 'var(--alert-color)' : 'var(--ha-color)'; ]]]
    notification: >
      [[[ @partial systeem
          return geheugen; ]]]
    notification_color: >
      [[[ @partial systeem
          return geheugen > 30 ? 'var(--alert-color)' : 'var(--ok-color)']]]
    notification_text_color: ivory
    info: >
      [[[ state_attr('update.home_assistant_core_update','installed_version'); ]]]
    info_color: >
      [[[ @partial systeem
          return update ? 'var(--alert-color)' : 'var(--ha-color)'; ]]]
```

keeps the colors

**Reply by elchininet** [reply]
You are using it right and your same code works on my side. Could you clean cache and check what styles are applied to your element?

**Reply by elchininet** [reply]
Select the anchor of the element, and you should see something like this:

![image](https://github.com/user-attachments/assets/ab9fc547-09e3-49a8-8a3d-a3d61261b8e8)

**Reply by Mariusthvdb** [reply]
cache is cleaned every time, and I can see that because of the working card_mod changes too.

let me test the styles


seems ok indeed:

<img width="985" alt="Scherm­afbeelding 2024-11-14 om 23 22 37" src="https://github.com/user-attachments/assets/00d64307-6209-4ed2-9281-57cd200b27f0">

but it shows like:

<img width="292" alt="Scherm­afbeelding 2024-11-14 om 23 23 16" src="https://github.com/user-attachments/assets/3986328f-5a5b-4b60-86b2-9b76bbc19599">



this is with the local keep:

<img width="958" alt="Scherm­afbeelding 2024-11-14 om 23 24 55" src="https://github.com/user-attachments/assets/0f55db02-2dd1-4906-81a5-35d9c09723cc">

**Reply by elchininet** [reply]
>seems ok indeed

No, it is not OK, I don't see the `selected` ones. The `selected` ones are OK in your second screenshot that is using item options and they should be in the first too if you have the global options (just like in my screenshot).

**Reply by elchininet** [reply]
Set the global options and select **the selected anchor**, if it doesn't have the `selected` variables, then it should not be the latest JavaScript code.

**Reply by elchininet** [reply]
Oh, wait I am seeing something in your screenshot: Safari. I always forget that you use Safari and it is the new Internet Explorer, let me test it in that browser.

**Reply by elchininet** [reply]
No, it is not that.
It is working properly in Safari too.

**Reply by elchininet** [reply]
In Safari go to the dev tools, go to Network and you will see in the right top corner a checkbox that says "disable cache", check it.

**Reply by Mariusthvdb** [reply]
so I did, and it seems to work now.


how odd. these must have been very persistent, because I clear cache all the time, and it always works....

closing this, because fixed.
thx!

**Reply by elchininet** [reply]
Sometimes the caching is very aggressive, best to keep that checkbox on and refresh the browser while you have the dev-tools open. That is a good way (not infalible though)
